File: 0740-delete-and-earn/0740-delete-and-earn.cpp
class Solution {
public:
    unordered_map<int, int> points;
    int deleteAndEarn(vector<int>& nums) {
        /*  - could we convert it to house ropper problem ? Yes
            - How ?
            sorted nums = 2 2 3 3 3 4
            unique = 2 3 4
            points 2 3 4
                   4 9 4

            1 3 5
        */
        sort(nums.begin(), nums.end());
        vector<int> unique;
        unique.push_back(nums[0]);
        for (int num : nums) {
            points[num] += num;
            if (num != unique.back()) {
                unique.push_back(num);
            }
        }

        return solve(unique, 0);
    }

    int solve(vector<int>& unique, int index) {
        if (index >= unique.size()) {
            return 0;
        }

        int take = 0, skip = 0;

        if (index < unique.size() - 1 && unique[index] == unique[index + 1] - 1)
            take = points[unique[index]] + solve(unique, index + 2);
        else
            take = points[unique[index]] + solve(unique, index + 1);

        skip = solve(unique, index + 1);
        return max(take, skip);
    }
};

File: 0740-delete-and-earn/0740-delete-and-earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:

        freq = Counter(nums)
        arr = list(freq.keys())
        arr.sort()
        memo = {}

        def solve(idx, prevPicked):
            if idx == len(arr):
                return 0
            key = (idx, prevPicked)

            if key in memo:
                return memo[key]

            take = skip = 0

            if not prevPicked or arr[idx] != (arr[idx - 1] + 1):
                val = arr[idx] * freq[arr[idx]]
                take = val + solve(idx + 1, True)

            skip = solve(idx + 1, False)

            memo[key] = max(take, skip)
            return max(take, skip)

        return solve(0, False)
